### PR TITLE
Lessons from porting mpc_follower

### PR DESCRIPTION
**Pull request by fred-apex-ai**
Extract the gist of the porting and save it for the next round

I didn't add updates on dynamic reconfigure to avoid conflicts with @TakaHoribe's #78

### CODE REVIEW

**Comment by mitsudome-r** [comment]
@fred-apex-ai Please also add documents about dynamic_reconfigure from this PR as well. https://github.com/tier4/AutowareArchitectureProposal/pull/78

File: docs/porting-to-ROS2.md [hunk 1]
```diff
@@ -45,10 +45,67 @@ A good general strategy is to try to implement those changes, then iteratively r
...
+
+```cmake
+if(NOT CMAKE_CXX_STANDARD)
+  set(CMAKE_CXX_STANDARD 14)
+  set(CMAKE_CXX_STANDARD_REQUIRED ON)
+  set(CMAKE_CXX_EXTENSIONS OFF)
+endif()
+```
+
+#### compiler flags
```
Comment (nnmm) [inline docs/porting-to-ROS2.md]: Nit: Uppercase, "headers" as well

File: docs/porting-to-ROS2.md [hunk 2]
```diff
@@ -45,10 +45,67 @@ A good general strategy is to try to implement those changes, then iteratively r
...
+Make sure that flags are only added explicitly that a compiler understands. Not everyone is using `gcc` or `clang`
```
Comment (nnmm) [inline docs/porting-to-ROS2.md]: Would reword as "Make sure that only those flags that a compiler understands are added explicitly", and add a period after the second sentence

File: docs/porting-to-ROS2.md [hunk 3]
```diff
@@ -45,10 +45,67 @@ A good general strategy is to try to implement those changes, then iteratively r
...
+
+```cmake
+if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+  add_compile_options(-Wall -Wextra -Wpedantic)
```
Comment (nnmm) [inline docs/porting-to-ROS2.md]: What about `-Wno-unused-parameter`?

File: docs/porting-to-ROS2.md [hunk 4]
```diff
@@ -45,10 +45,67 @@ A good general strategy is to try to implement those changes, then iteratively r
...
+endif()
+```
+
+#### headers
+It is good practice -- though not required for the build to succeed -- to add all source files explicity in `cmake` so e.g. IDE know if a file belongs to the project or not. For example,
```
Comment (nnmm) [inline docs/porting-to-ROS2.md]: That goes against the ROS recommendations: "only cpp files are explicitly referenced in the call to add_library or add_executable". Does the IDE find the headers as well when they are added with 
```
target_include_directories(my_target
  PUBLIC
    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
    $<INSTALL_INTERFACE:include>)
```
like in the ROS recommendations?

File: docs/porting-to-ROS2.md [hunk 5]
```diff
@@ -130,6 +187,10 @@ which is equivalent to
 
     const double vel_lim = declare_parameter<double>("vel_lim", 25.0);
 
+This makes the parameter visible to ros2 and its initial value can be set e.g. via a parameter file.
```
Comment (nnmm) [inline docs/porting-to-ROS2.md]: I'm not sure what "This makes the parameter visible to ros2" means – that without this, trying to set the parameter will be ignored? Maybe this could be a bit more explicit.

File: docs/porting-to-ROS2.md [hunk 6]
```diff
@@ -138,6 +199,19 @@ Two levels of hierarchy need to be added around the parameters themselves:
       ros__parameters:
         <params>
 
+
+##### Indendation
+Without proper indentation (4 spaces in the example above), there is a segfault when the YAML is
+parsed during `rclcpp::init(argc, argv)`:
```
Comment (nnmm) [inline docs/porting-to-ROS2.md]: Maybe put this part in the section about confusing error messages below, and only say here that correct indentation is important?

File: docs/porting-to-ROS2.md [hunk 7]
```diff
@@ -174,14 +248,59 @@ You could do therefore try setting up a dedicated thread, but you could also use
...
     // In the function processing data
-    tf_buffer_.waitForTransform(a, b, msg_time, std::chrono::milliseconds(1000), [this](const std::shared_future<geometry_msgs::msg::TransformStamped>& tf){
+    auto tf_future = tf_buffer_.waitForTransform(a, b, msg_time, std::chrono::milliseconds(1000), [this](const std::shared_future<geometry_msgs::msg::TransformStamped>& tf){
         // Here the future is available
     });
 
 The callback will always be called, but only after some time: when the transform becomes available or when the timeout is reached. In the latter case, if the transform is not ready yet, calling `.get()` on the future will throw a `tf2::TimeoutException`.
 
-The `waitForTransform()` function will return immediately and also return a future. However, calling `.get()` or `.wait()` on that future does not respect the timeout. That is, it will wait however long it takes until a transform arrives and never throw an exception.
+But there is a better way that doesn't need exceptions. The `waitForTransform()` function will
```
Comment (nnmm) [inline docs/porting-to-ROS2.md]: I disagree that this is better – see the discussion on my previous PR to this doc, or we can hop on a call :)

File: docs/porting-to-ROS2.md [hunk 8]
```diff
@@ -174,14 +248,59 @@ You could do therefore try setting up a dedicated thread, but you could also use
...
+      RCLCPP_INFO(
+        get_logger(), "waiting another %d seconds for %s->%s transform",
+        std::chrono::duration_cast<std::chrono::seconds>(timeout).count(), input, output);
+    }
+  }
+}
+
+```
 
+### Shared pointers
```
Comment (nnmm) [inline docs/porting-to-ROS2.md:390]: Does this scenario tend to come up a lot when porting? If yes, where?

File: docs/porting-to-ROS2.md [hunk 9]
```diff
@@ -201,15 +320,27 @@ Another idea for a workaround is to do something similar to what is done in the
 
 
 ### Logging
-The node name is now automatically prepended to the log message, so that part can be removed.
+The node name is now automatically prepended to the log message, so that part can be removed. Get the logger from the node, and if necessary, update free functions to accept a logger as a new argument; e.g.,
+
+    ROS_INFO_COND(show_debug_info_, "[MPC] some message with a float value %g", some_member_);
 
+should become
```
Comment (nnmm) [inline docs/porting-to-ROS2.md:423]: :+1:

File: docs/porting-to-ROS2.md (hunk 4 above)
Comment (fred-apex-ai) [reply docs/porting-to-ROS2.md]: Turns out at least CLion is clever enough, and we don't want to make the `CMakeLists.txt` harder to read, so I removed the entire section

File: docs/porting-to-ROS2.md (hunk 6 above)
Comment (fred-apex-ai) [reply docs/porting-to-ROS2.md]: yeah, I hadn't seen that section at the bottom. You got the same error  message for a different reason (why not?)

File: docs/porting-to-ROS2.md (hunk 8 above)
Comment (fred-apex-ai) [reply docs/porting-to-ROS2.md:390]: I don't know if it happens often but it happened to me once [here](https://github.com/tier4/Pilot.Auto/pull/52/files?file-filters%5B%5D=.cfg&file-filters%5B%5D=.cpp&file-filters%5B%5D=.py&file-filters%5B%5D=.test&file-filters%5B%5D=.txt&file-filters%5B%5D=.xml&file-filters%5B%5D=.yaml&file-filters%5B%5D=No+extension#r510717033) and every instance is one too many

File: docs/porting-to-ROS2.md (hunk 7 above)
Comment (fred-apex-ai) [reply docs/porting-to-ROS2.md]: @nnmm yes, that's a tough one. perhaps I just leave my changes out, we figure out the proper way to go forward in #103 and then update this doc later.

File: docs/porting-to-ROS2.md [hunk 10]
```diff
@@ -391,7 +499,7 @@ If you forget `<build_type>ament_cmake</build_type>`, or you use package format
 
 
 ### YAML param file
-Used tabs instead of spaces in your param.yaml file? _Clearly_, the most user-friendly error message is
+Used tabs instead of spaces in your param.yaml file? Or didn't indent properly? _Clearly_, the most user-friendly error message is
```
Comment (nnmm) [inline docs/porting-to-ROS2.md]: Did you check that it's the same error message? This one doesn't mention "double free detected in tcache 2", so maybe it's better to follow the error message with "a similar message occurs when indentation is inconsistent, but it says 'double free detected in tcache 2'" instead

File: docs/porting-to-ROS2.md (hunk 7 above)
Comment (fred-apex-ai) [reply docs/porting-to-ROS2.md]: @nnmm and I discussed this on the whiteboard. There are essentially two use cases

## node ready?
The goal is to ensure that node can do its job properly and that the transforms it needs are generally available.
There are two options we have
1. wait indefinitely in the constructor
2. check locally where the tf is to be used if it's available. None of that requires `waitForTransform`

I take over the responsibility and write down the recomendation in a separate PR

## getting the right transform
For example if latencies in the system are an issue and one node has to wait for the transform to arrive for a particular timestamp to best transform a data set

This part still requires some investigation and I leave it to @nnmm to carry out the investigation and report the conclusion in this porting guidelines in a separate PR

**Comment by fred-apex-ai** [comment]
@nnmm SOrry for taking so long, please a final look. Ready from my side, resolved all discussions with new commits

File: docs/porting-to-ROS2.md [hunk 11]
```diff
@@ -130,14 +170,181 @@ which is equivalent to
...
+
+Add a method to declare all the parameters
+
+    void declareMPCparameters();
+
+#### implementation file
+
+Write the following into the definition of the class that inherits from `rclcpp::Node`.
+
+A few macro and a utility function help keep the following code and void of redundancy
```
Comment (nnmm) [inline docs/porting-to-ROS2.md]: macros plural? Also, missing word after "code"

Comment (nnmm) [reply docs/porting-to-ROS2.md]: Also, mention somewhere that "MPC" is just the name of the example?

Comment (nnmm) [reply docs/porting-to-ROS2.md]: And perhaps mention that these are entirely optional/make sense when there is a lot of repetition. I think macros are discouraged by the Google Style guide at least.

File: docs/porting-to-ROS2.md [hunk 12]
```diff
@@ -130,14 +170,181 @@ which is equivalent to
...
+
+```
+$ ros2 param set /mpc_follower prediction_horizon 51
+```
+
+Make sure relevant parameters can be set from the command line or `rqt` and changes are reflected by
+the package.
+
+#### Parameter client [discouraged]
+The parameter client is another way to dynamically set the parameter defined in the node. The client subscribes to the `/parameter_event` topic and call the callback function. This allows the client node to get all the information about parameter change in every node. The callback argument contains the target node name, which can be used to determine which node the parameter change is for.
```
Comment (nnmm) [inline docs/porting-to-ROS2.md]: Nit: plural "about parameter change**s**"?

File: docs/porting-to-ROS2.md [hunk 13]
```diff
@@ -146,6 +353,7 @@ Also, ROS1 didn't have a problem when you specify an integer, e.g. `28` for a `d
 Best to just change `28` to `28.0` in the param file. See also [this issue](https://github.com/ros2/rclcpp/issues/979).
 
 
+
```
Comment (nnmm) [inline docs/porting-to-ROS2.md]: Nit: Let's keep two newlines between sections

File: docs/porting-to-ROS2.md [hunk 14]
```diff
@@ -159,8 +367,8 @@ However, in some cases the extra functionality of `tf2_ros::Buffer` is needed. F
 #### Waiting for a transform to arrive
 You might expect to be able to use `tf2_ros::Buffer::lookupTransform()` with a timeout out of the box, but this will throw an error:
 
-    Do not call canTransform or lookupTransform with a timeout unless you are using another thread for populating data. 
-    Without a dedicated thread it will always timeout. 
+    Do not call canTransform or lookupTransform with a timeout unless you are using another thread for populating data.
```
Comment (nnmm) [inline docs/porting-to-ROS2.md:367]: Nit: Code formatting for the functions

File: docs/porting-to-ROS2.md [hunk 15]
```diff
@@ -174,14 +382,24 @@ You could do therefore try setting up a dedicated thread, but you could also use
...
     // In the function processing data
-    tf_buffer_.waitForTransform(a, b, msg_time, std::chrono::milliseconds(1000), [this](const std::shared_future<geometry_msgs::msg::TransformStamped>& tf){
+    auto tf_future = tf_buffer_.waitForTransform(a, b, msg_time, std::chrono::milliseconds(1000), [this](const std::shared_future<geometry_msgs::msg::TransformStamped>& tf){
         // Here the future is available
     });
 
 The callback will always be called, but only after some time: when the transform becomes available or when the timeout is reached. In the latter case, if the transform is not ready yet, calling `.get()` on the future will throw a `tf2::TimeoutException`.
 
-The `waitForTransform()` function will return immediately and also return a future. However, calling `.get()` or `.wait()` on that future does not respect the timeout. That is, it will wait however long it takes until a transform arrives and never throw an exception.
+The `waitForTransform()` function will return immediately and also return a future. However, calling
```
Comment (nnmm) [inline docs/porting-to-ROS2.md]: Nit: Please remove mid-sentence linebreaks

File: docs/porting-to-ROS2.md (hunk 8 above)
Comment (nnmm) [reply docs/porting-to-ROS2.md:390]: I was just wondering whether it's a general C++ tip or specific to ROS1->ROS2 porting. I'll leave it up to you.

File: docs/porting-to-ROS2.md [hunk 16]
```diff
@@ -285,3 +536,21 @@ Used tabs instead of spaces in your param.yaml file? _Clearly_, the most user-fr
     [ERROR] [mypackage-1]: process has died [pid 30427, exit code -6, cmd '/home/user/workspace/install/mypackage/lib/mypackage/mypackage --ros-args -r __node:=mypackage --params-file /home/user/workspace/install/mypackage/share/mypackage/param/myparameters.yaml'].
 
 and that is indeed what ROS2 will tell you.
+
+#### No indentation
+
+Without proper indentation of levels, there is a segfault when the YAML is parsed during `rclcpp::init(argc, argv)`. The
```
Comment (nnmm) [inline docs/porting-to-ROS2.md]: Nit: Remove mid-sentence linebreak

File: docs/porting-to-ROS2.md [hunk 17]
```diff
@@ -285,3 +536,21 @@ Used tabs instead of spaces in your param.yaml file? _Clearly_, the most user-fr
...
+error is similar to the above but begins with
+
+    [mpc_follower-1] free(): double free detected in tcache 2
+
+Note that this message may be hidden when just launching with `ros2 launch`. It is shown running the node under
```
Comment (nnmm) [inline docs/porting-to-ROS2.md]: Nit: Remove mid-sentence linebreak

**Review by nnmm — APPROVED** [review]
LGTM with a few nits

**Comment by fred-apex-ai** [comment]
@mitsudome-r please merge

**Comment by fred-apex-ai** [comment]
@TakaHoribe please merge

**Comment by TakaHoribe** [comment]
@fred-apex-ai Thank you!
